File: Languages/go/algorithms-dsa/leetcode/backtracking/n_queens.go
package main

import "fmt"

/*
  LC 51 - N-Queens
  Topic: Backtracking
  Difficulty: Hard
  Time: O(n!) | Space: O(n^2)
*/

func solveNQueens(n int) [][]string {
	var result [][]string
	board := make([][]byte, n)
	for i := range board {
		board[i] = make([]byte, n)
		for j := range board[i] {
			board[i][j] = '.'
		}
	}
	cols := make([]bool, n)
	diag1 := make([]bool, 2*n)
	diag2 := make([]bool, 2*n)

	var backtrack func(row int)
	backtrack = func(row int) {
		if row == n {
			sol := make([]string, n)
			for i := range board {
				sol[i] = string(board[i])
			}
			result = append(result, sol)
			return
		}
		for col := 0; col < n; col++ {
			if cols[col] || diag1[row-col+n] || diag2[row+col] {
				continue
			}
			board[row][col] = 'Q'
			cols[col], diag1[row-col+n], diag2[row+col] = true, true, true
			backtrack(row + 1)
			board[row][col] = '.'
			cols[col], diag1[row-col+n], diag2[row+col] = false, false, false
		}
	}
	backtrack(0)
	return result
}

func main() {
	res := solveNQueens(4)
	if len(res) == 2 {
		fmt.Println("PASS: solveNQueens(4) = 2 solutions")
	} else {
		fmt.Println("FAIL: solveNQueens(4), got", len(res))
	}

	res1 := solveNQueens(1)
	if len(res1) == 1 {
		fmt.Println("PASS: solveNQueens(1) = 1 solution")
	} else {
		fmt.Println("FAIL: solveNQueens(1)")
	}
}

File: Languages/python/algorithms-dsa/leetcode/backtracking/n_queens.py
"""
LeetCode #51 - N-Queens
Topic: Backtracking
Difficulty: Hard

Place N queens on an N×N board row-by-row with column/diagonal tracking.

Time Complexity: O(n!)
Space Complexity: O(n)
"""


def solve_n_queens(n: int) -> list[list[str]]:
    result: list[list[str]] = []
    cols: set[int] = set()
    diag1: set[int] = set()  # row - col
    diag2: set[int] = set()  # row + col
    board = [["." for _ in range(n)] for _ in range(n)]

    def backtrack(row: int) -> None:
        if row == n:
            result.append(["".join(r) for r in board])
            return
        for col in range(n):
            if col in cols or (row - col) in diag1 or (row + col) in diag2:
                continue
            cols.add(col)
            diag1.add(row - col)
            diag2.add(row + col)
            board[row][col] = "Q"
            backtrack(row + 1)
            board[row][col] = "."
            cols.remove(col)
            diag1.remove(row - col)
            diag2.remove(row + col)

    backtrack(0)
    return result


if __name__ == "__main__":
    assert len(solve_n_queens(4)) == 2
    assert len(solve_n_queens(1)) == 1
    assert solve_n_queens(1) == [["Q"]]
    print("All tests passed!")
